src/Git.hs

```diff
@@ -201,100 +199,99 @@ instance FromJSON GitIntegrationFailure
-  | ForcePush Sha Branch (PushResult -> a)
-  | Push Sha Branch (PushResult -> a)
-  | PushAtomic [SomeRefSpec] (PushResult -> a)
-  | Rebase Sha RemoteBranch (Maybe Sha -> a)
-  | Merge Sha Text (Maybe Sha -> a)
-  | Checkout RemoteBranch (Maybe Sha -> a)
-  | Clone RemoteUrl (CloneResult -> a)
-  | GetParent Sha (Maybe Sha -> a)
-  | RevParse SomeRefSpec (Maybe Sha -> a)
-  | DoesGitDirectoryExist (Bool -> a)
-  | LastTag Sha (Maybe Text -> a)
-  | ShortLog SomeRefSpec SomeRefSpec (Maybe Text -> a)
-  | Tag Sha TagName TagMessage (TagResult -> a)
-  | DeleteTag TagName a
-  | DeleteBranch Branch a
-  | DeleteRemoteBranch Branch (PushResult -> a)
-  | CheckOrphanFixups Sha RemoteBranch (Bool -> a)
-  deriving (Functor)
-
-type GitOperation = Free GitOperationFree
-
-fetchBranch :: Branch -> GitOperation ()
-fetchBranch remoteBranch = liftF $ FetchBranch remoteBranch NoTags ()
-
-fetchBranchWithTags :: Branch -> GitOperation ()
-fetchBranchWithTags remoteBranch = liftF $ FetchBranch remoteBranch WithTags ()
-
-forcePush :: Sha -> Branch -> GitOperation PushResult
-forcePush sha remoteBranch = liftF $ ForcePush sha remoteBranch id
-
-push :: Sha -> Branch -> GitOperation PushResult
-push sha remoteBranch = liftF $ Push sha remoteBranch id
-
-pushAtomic :: [SomeRefSpec] -> GitOperation PushResult
-pushAtomic refs = liftF $ PushAtomic refs id
-
-rebase :: Sha -> RemoteBranch -> GitOperation (Maybe Sha)
-rebase sha ontoBranch = liftF $ Rebase sha ontoBranch id
-
-merge :: Sha -> Text -> GitOperation (Maybe Sha)
-merge sha message = liftF $ Merge sha message id
+data GitOperation :: Effect where
+  FetchBranch :: Branch -> FetchWithTags -> GitOperation m ()
+  ForcePush :: Sha -> Branch -> GitOperation m PushResult
+  Push :: Sha -> Branch -> GitOperation m PushResult
+  PushAtomic :: [SomeRefSpec] -> GitOperation m PushResult
+  Rebase :: Sha -> RemoteBranch -> GitOperation m (Maybe Sha)
+  Merge :: Sha -> Text -> GitOperation m (Maybe Sha)
+  Checkout :: RemoteBranch -> GitOperation m (Maybe Sha)
+  Clone :: RemoteUrl -> GitOperation m CloneResult
+  GetParent :: Sha -> GitOperation m (Maybe Sha)
+  RevParse :: SomeRefSpec -> GitOperation m (Maybe Sha)
+  DoesGitDirectoryExist :: GitOperation m Bool
+  LastTag :: Sha -> GitOperation m (Maybe Text)
+  ShortLog :: SomeRefSpec -> SomeRefSpec -> GitOperation m (Maybe Text)
+  Tag :: Sha -> TagName -> TagMessage -> GitOperation m TagResult
+  DeleteTag :: TagName -> GitOperation m ()
+  DeleteBranch :: Branch -> GitOperation m ()
+  DeleteRemoteBranch :: Branch -> GitOperation m PushResult
+  CheckOrphanFixups :: Sha -> RemoteBranch -> GitOperation m Bool
+
+type instance DispatchOf GitOperation = 'Dynamic
+
+fetchBranch :: GitOperation :> es => Branch -> Eff es ()
+fetchBranch remoteBranch = send $ FetchBranch remoteBranch NoTags
+
+fetchBranchWithTags :: GitOperation :> es => Branch -> Eff es ()
+fetchBranchWithTags remoteBranch = send $ FetchBranch remoteBranch WithTags
```

Comment:
Here you can see the difference in jargon. These functions are mechanical rewrites of the `free` functions. If you've seen one of them, you've seen them all.

So instead of running `liftF` on an action, you `send` it, and you don't pass dummy continuations like `id` and `()`.

Reply:
Good question! I did consider it, but the following made me decide against it:

- I didn't come across any other template haskell, this would introduce it
- TH comes with quite a few downsides, which could make this PR more controversial
- in a personal project I did initially use effectful-th, but was forced to remove it again after the staticly linked build failed. Judging by the error it likely was more of a Nix/ghc problem than an effectful-th problem, but it does add to the reasons for not using it
- with the educative goal of this PR it might be better to write the definitions out in full.

Reply:
Yeah I agree that not adding TH splices here if there aren't any already is probably a good call!

Reply:
Yeah, we talked a bit about that as well, and I too think we're better off without `TH` in this case.